pkg/webhooks/clusterlogging/clusterlogging.go

```diff
@@ -0,0 +1,250 @@
+// Ignore means that an error calling the webhook is ignored and the API request is allowed to continue.
+// It's important to leave the FailurePolicy set to Ignore because otherwise the pod will fail to be created as the API request will be rejected.
+func (s *ClusterloggingWebhook) FailurePolicy() admissionregv1.FailurePolicyType {
+	return admissionregv1.Ignore
+}
+
+// Rules implements Webhook interface
+func (s *ClusterloggingWebhook) Rules() []admissionregv1.RuleWithOperations { return rules }
+
+// GetURI implements Webhook interface
+func (s *ClusterloggingWebhook) GetURI() string { return "/" + WebhookName }
+
+// SideEffects implements Webhook interface
+func (s *ClusterloggingWebhook) SideEffects() admissionregv1.SideEffectClass {
+	return admissionregv1.SideEffectClassNone
+}
+
+// Validate implements Webhook interface
+func (s *ClusterloggingWebhook) Validate(req admissionctl.Request) bool {
+	valid := true
+	valid = valid && (req.UserInfo.Username != "")
+	valid = valid && (req.Kind.Kind == ClusterLoggingKind)
+
+	return valid
+}
+
+type retentionPolicyValidator struct {
+	name       string
+	hint       string
+	upperBound esapi.TimeUnit
+}
+
+func (r *retentionPolicyValidator) checkPolicy(retentionPolicy *cl.RetentionPolicySpec) (bool, admissionctl.Response) {
+	isAllowed, deniedMessage, err := r.isAllowed(retentionPolicy)
+	if err != nil {
+		return false, admissionctl.Errored(http.StatusBadRequest, err)
+	}
+	if !isAllowed {
+		return false, admissionctl.Denied(deniedMessage)
+	}
+	return true, admissionctl.Allowed("Allowed to create ClusterLogging")
+}
+
+func (r *retentionPolicyValidator) isAllowed(retentionPolicy *cl.RetentionPolicySpec) (bool, string, error) {
+	if retentionPolicy == nil {
+		return false, "The entered retention policy is not allowed. " + r.name + " must not be unset. Hint: " + r.hint, nil
+	}
+
+	isAllowedRetentionDays, err := le(retentionPolicy.MaxAge, r.upperBound)
+	if err != nil {
+		log.Error(err, "Couldn't compare timeunits")
+		return false, "", err
+	}
+
+	if !isAllowedRetentionDays {
+		return false, "The entered RetentionPolicy " + r.name + " is not allowed. " + r.hint, nil
+	}
+
+	return true, "", nil
+}
+
+func (s *ClusterloggingWebhook) Authorized(request admissionctl.Request) admissionctl.Response {
+	r := s.authorized(request)
+	r.UID = request.AdmissionRequest.UID
+	return r
+}
+
+// Authorized implements Webhook interface
+func (s *ClusterloggingWebhook) authorized(request admissionctl.Request) admissionctl.Response {
+	var ret admissionctl.Response
+
+	decoder, err := admissionctl.NewDecoder(&s.s)
+	if err != nil {
+		return admissionctl.Errored(http.StatusBadRequest, err)
+	}
+	clusterLogging := &cl.ClusterLogging{}
+	if len(request.OldObject.Raw) > 0 {
+		err = decoder.DecodeRaw(request.OldObject, clusterLogging)
+	} else {
+		err = decoder.DecodeRaw(request.Object, clusterLogging)
+	}
```

Comment:
```suggestion
// renderClusterLogging decodes an *cl.ClusterLogging from the incoming request.
// If the request includes an OldObject (from an update or deletion), it will be
// preferred, otherwise, the Object will be preferred.
func (s *ClusterloggingWebhook) renderClusterLogging(request admissionctl.Request) (*cl.ClusterLogging, error) {
	decoder, err := admissionctl.NewDecoder(&s.s)
	if err != nil {
		return nil, err
	}
	clusterLogging := &cl.ClusterLogging{}
	if len(request.OldObject.Raw) > 0 {
		err = decoder.DecodeRaw(request.OldObject, clusterLogging)
	} else {
		err = decoder.DecodeRaw(request.Object, clusterLogging)
	}
	if err != nil {
		return nil, err
	}
	return clusterLogging, nil
}

// Authorized implements Webhook interface
func (s *ClusterloggingWebhook) authorized(request admissionctl.Request) admissionctl.Response {
	var ret admissionctl.Response

	clusterLogging, err := s.renderClusterLogging(request)
	if err != nil {
		return admissionctl.Errored(http.StatusBadRequest, err)
	}
```

In other places where objects are decoded, we have the convention to put those operations into their own method ([cf namespace hook](https://github.com/openshift/managed-cluster-validating-webhooks/blob/3031ecc55e118a436db367fd153025d396d01799/pkg/webhooks/namespace/namespace.go#L111-L140)), so I've made a Github code suggestion to do that here. Please have a look and let me know what you think.